### ISSUE TITLE
"Feature Request: Extend random_circuit function to support gate number and gate set specification"

### ISSUE THREAD

**Issue by PhysicsQoo** [Qiskit/qiskit#13053]
### What should we add?

I would like to request an enhancement to the random_circuit function in qiskit/circuit/random/utils.py. Currently, the function allows specifying only the depth of the circuit, but it would be highly useful to have additional parameters to:

- Specify the number of gates in the circuit.
- Specify a custom set of gates to be used, similar to the functionality provided by the random_clifford_circuit function.

This enhancement would allow users to have more control over the structure of the randomly generated circuit, which is particularly useful for research and testing purposes where specific gate constraints are required.

- Add an optional parameter num_gates to specify the exact number of gates in the generated circuit.
- Add an optional parameter gates to allow users to provide a list of gate types that should be randomly selected during circuit generation.

This feature would make random_circuit more versatile and better suited for a wider range of applications.
It would provide more flexibility for researchers who need to simulate or analyze circuits with specific gate constraints.

### Reference

1. [random_circuit function documentation](https://docs.quantum.ibm.com/api/qiskit/circuit#random_circuit)
2. [random_clifford_circuit function documentation](https://docs.quantum.ibm.com/api/qiskit/circuit#random_clifford_circuit)
3. [Qiskit Random Circuit Repository](https://github.com/Qiskit/qiskit/blob/cc2edc9f7172d75ec454fdaf5748326cd3720f90/qiskit/circuit/random/utils.py#L24)

**Comment by Cryoris** [comment]
Specifying the gates sounds like a good idea. The `random_circuit` function should likely also use the `qiskit.circuit.library.standard_gates.get_standard_gate_name_mapping` to get the gates, instead of keeping a separate hardcoded list. I see you've already opened a first PR (thanks for that!) but maybe we can tackle the problems in a different fashion and resolve the problems together.

I would suggest we add a `gates: list[str] | None` argument to `random_circuit`, which allows selecting the gates (this would make `random_clifford` and `random_clifford_t` one-liners). We can use the `get_standard_gate_mapping` to determine the list of gates, which returns a dictionary of `{name: gate_instance}` pairs. What do you think? Let me know if you'd like some pointers on how to do that 🙂

**Comment by PhysicsQoo** [comment]
I agree that adding a `gates: list[str] | None` argument is a great idea, as it would allow users to easily select their desired gate set. This would indeed make the `random_clifford` and `random_clifford_t` functions much simpler.

Additionally, while the `random_circuit` function currently uses `depth` to determine the circuit size, I think it would be beneficial to also have an option to specify the number of gates, similar to the `random_clifford` function. Do you think it would be better to retain the `depth` argument alongside a new `num_gates` option, replace `depth` with `num_gates`, or perhaps support both options simultaneously? I’m interested in your thoughts on the best approach.
